[clang][modules-driver] Add dependency scan and dependency graph

naveen-seth opened the pull request:
This patch is part of a series to support driver-managed module builds. 
It adds support for discovery of module dependencies from within the driver and for generation of the module dependency graph.

Source inputs provided on the command line can import modules defined in the Standard library module manifest, which are then scanned on demand.

The dependency scan and graph support both Clang modules and C++ named modules.
The generated dependency graph can be output in Graphviz DOT format as a remark.

<img width="450" alt="image" src="https://github.com/user-attachments/assets/b144e93d-1d3f-4f6a-9b68-aa541462c812" /><br>

RFC discussing linking the driver against additional libraries:
https://discourse.llvm.org/t/rfc-driver-link-the-driver-against-clangdependencyscanning-clangast-clangfrontend-clangserialization-and-clanglex

RFC for driver-managed module builds:
https://discourse.llvm.org/t/rfc-modules-support-simple-c-20-modules-use-from-the-clang-driver-without-a-build-system

naveen-seth commented on clang/lib/Driver/ModulesDriver.cpp:
```diff
@@ -0,0 +1,1579 @@
+  Inputs.emplace_back(types::TY_CXXModule, A);
+}
+
+void buildStdModuleManifestInputs(const StdModuleManifest &Manifest,
```
Currently, this allows for importing any module defined in the standard module manifest, even if the manifest contains `is-std-library: false` for that module entry.
Should we constrain this?

ChuanqiXu9 replied:
What is the status quo of distributed different std manifest files ?

naveen-seth replied:
Done. (And also thank you for reviewing this.)

ChuanqiXu9 commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -1829,6 +1831,53 @@ Compilation *Driver::BuildCompilation(ArrayRef<const char *> ArgList) {
+    // pruned later.
+    const auto StdModuleManifestPath =
+        GetStdModuleManifestPath(*C, C->getDefaultToolChain());
+    if (StdModuleManifestPath == "<NOT PRESENT>") {
```
nit: maybe we can avoid hardcoding by judging this by if it is a path.

naveen-seth replied:
done.

ChuanqiXu9 commented on clang/lib/Driver/ModulesDriver.cpp:
```diff
@@ -0,0 +1,1579 @@
+static bool hasCXXNamedModuleInput(const InputList &Inputs) {
+  const auto IsTypeCXXModule = [](const auto &Input) -> bool {
+    const auto TypeID = Input.first;
+    return (TypeID == types::TY_CXXModule);
```
Should we judge preprocessed CXX module for TY_PP_CXXModule?

ChuanqiXu9 commented on clang/lib/Driver/ModulesDriver.cpp:
```diff
@@ -0,0 +1,1579 @@
+    FixIts.emplace_back(SrcMgr, LangOpts, FixIt);
+}
+
+/// Translates \c StandaloneDiag into a StoredDiagnostic, associating it with
```
Can we explain why we need it?

ChuanqiXu9 commented on clang/lib/Driver/ModulesDriver.cpp:
```diff
@@ -0,0 +1,1579 @@
+
+  private:
+    /// We need the output of dependency scan to be deterministic. However,
+    /// the dependency graph may contain two modules with the same name. How
```
Then it is an error.

Bigcheese replied:
Since this is a merged graph from multiple inputs, multiple modules with the same name is not an error. When doing multi-arch builds, or offloading you need different PCMs for each target arch.

Bigcheese replied:
This can happen with named modules too, although the current code probably only works for the multi-arch case.

While the current code doesn't support it, I would like to eventually support:
`clang++ file.cpp file.mm module.cppm -o program`
Where both file.cpp and file.mm import module.cppm. This requires two different PCMs and one .o file for the module.

naveen-seth replied:
For named modules, we currently signal an error here, while building the graph:
https://github.com/naveen-seth/llvm-project/blob/f1d92657d1d47bd1f436546d823bc75757e1443a/clang/lib/Driver/ModulesDriver.cpp#L1043-L1046
@ChuanqiXu9.

ChuanqiXu9 commented on clang/lib/Driver/ModulesDriver.cpp:
```diff
@@ -0,0 +1,1579 @@
+    SmallVector<std::unique_ptr<ModuleDeps>> takeOrderedModuleDeps();
+
+  private:
+    /// We need the output of dependency scan to be deterministic. However,
```
If I read correctly, let's add a `FIXME: ` here

ChuanqiXu9 commented:
BTW, if this is not too late, I feel this work might be worth to be in other clang module (maybe in supporting) than driver (and let driver to link it). The idea is, the tools needs to understand modules, but it is complex. You can find similar works in clangd's ModulesBuilder. It is too expensive for every tool (and clang-tidy or whatever else) to do the same thing. It will be good if clang can provide such ability for it. And I think your work is a good candidate.

In one word, I hope this to be ability of clang to allow clang based tools to understand modules (of course, including the clang driver itself).

naveen-seth commented on clang/include/clang/Driver/ModulesDriver.h:
```diff
@@ -0,0 +1,112 @@
+/// Identical to Driver::InputList.
+using InputList = llvm::SmallVector<InputTy, 16>;
+
+/// Checks whether the -fmodules-driver feature should be implicitly enabled.
```
Currently, the logic to implicitly enable the feature is only used for diagnostics but doesn't actually do anything. 

https://github.com/llvm/llvm-project/blob/cb0ac589ecc742ae6933a634c577eeb46831114d/clang/lib/Driver/Driver.cpp#L1838-L1841

This was already introduced in #153497 and is just being relocated in this patch. Since this patch introduces the `ModulesDriver.cpp` file, I thought it made sense to also move as much modules-driver–related code there as possible.

naveen-seth replied:
To confirm, you also want to revert #153497 for now?

Bigcheese commented:
> [...] But if we are able to make this to be a foundation for all clang based tools to understand modules. This will be pretty valuable.

This was something I was considering during the initial proposal, but I didn't suggest doing it that way because I don't think we know what that API should look like yet, and there are still open questions about how to handle 3rd party modules and header units. I think the exercise of making all this work in the constrained environment of the Clang driver will help answer what the eventual API should be.

With that, there are definitely parts of ModuleDriver.cpp that could be moved to the DependencyScanning library, and this library will soon move out of tooling into its own top level library. This did make me notice an issue with this patch is that it likely will break shared library builds because DependencyScanning currently depends on Driver, and this makes Driver depend on DependencyScanning. I think this should be fixed by just making DependencyScanning not depend on the driver (I'll expand on that in another comment).

I think a few parts of ModuleDriver.cpp should land in DependencyScanning in this patch, and after the modules driver is working well see how best to combine that with what clangd and other tools need, but until then I'd like the modules driver to be flexible as we figure things out, moving shared things to DependencyScanning as needed.

Longer term I've been wondering if something like how the [new Swift driver](https://github.com/swiftlang/swift-driver) works would be interesting for Clang. The Swift driver can be directly used by build systems to figure out all the work needed to build arbitrary sets of code, including any Clang modules that are needed. It can also use its internal build system to build directly. This would be a big change for Clang in that it takes on a lot of build system responsibilities, but it has worked really well for Swift.

Bigcheese commented on clang/lib/Driver/CMakeLists.txt:
```diff
@@ -99,5 +100,6 @@ add_clang_library(clangDriver
   LINK_LIBS
   clangBasic
   clangLex
+  clangDependencyScanning
```
As mentioned in my previous comment, this causes a cycle because clangDependencyScanning depends on clangDriver. This is fine when statically linking, but will fail with dynamic linking which we support.

This is only used by `DependencyScanningWorker::scanDependencies` to handle driver commands. To fix this I think it would be best to remove this dependency and instead have `scanDependencies` only work on frontend commands. I would then add a function to `clangDriver` that takes a `DependencyScanningWorker` and runs the scanner on all of frontend jobs, this would then be used to replace calls to `scanDependencies`.

This will need to be handled before this PR lands to avoid build issues.

Bigcheese replied:
Frontend depends on Driver for the options (because driver and -cc1 options got merged) and for `clang::createInvocation`. I think we should move the options to their own library (clangOptions?). `clang::createInvocation` is harder because it's used by `ASTUnit::LoadFromCommandLine`. I think both of these should move to the driver as the rest of the frontend doesn't use them.

shafik commented on clang/include/clang/Basic/DiagnosticDriverKinds.td:
```diff
@@ -587,6 +587,24 @@ def remark_found_cxx20_module_usage : Remark<
   "performing driver managed module build">,
   InGroup<ModulesDriver>;
+def remark_modules_manifest_not_found : Remark<
+  "standard modules manifest file not found; import of standard library "
```
ping

naveen-seth committed:
[clang][modules-driver] Add dependency scan and dependency graph

This patch is part of a series to support driver-managed module builds.
It adds support for discovery of module dependencies from within the
driver and for generation of the module dependency graph.

Source inputs provided on the command line can import modules defined
in the standard module manifest, which are then scanned on demand.

The dependency scan and graph support both Clang modules and C++ named
modules, and work correctly with multi-architecture and offloading
command lines. The generated dependency graph can be output in
Graphviz DOT format as a remark.

naveen-seth commented:
This should be ready for review again.

The latest version of this code should now also support multi-arch/offloading compilations. 
For now, I’ve removed the offloading test case (0264fb2) because I’m not able to reliably restrict it to environments where we actually generate offloading jobs.
(I also can’t provide a multi-arch test at the moment since I don’t have access to a macOS toolchain)

naveen-seth committed:
Fixup docstring

naveen-seth committed:
Fix naming

jansvoboda11 reviewed (commented):
I skimmed the PR and it looks reasonable to me. I think you'll have easier time getting reviews if you split it up into smaller chunks. For example you could extract the small refactors in `Driver.h` and `Driver.cpp`, add the multi-threading and graph rendering later, etc.

naveen-seth committed:
Merge remote-tracking branch 'upstream/main' into modules-driver-scan-and-graph

naveen-seth committed:
Fix broken test

naveen-seth committed:
Undo accidental renaming

naveen-seth committed:
Fix formatting

naveen-seth committed:
Use thread-safe FS for scans

naveen-seth committed:
Fix misplaced docstring.

naveen-seth commented:
Ping

naveen-seth committed:
Merge branch 'main' into modules-driver-scan-and-graph

naveen-seth committed:
Fix build failure caused by updating branch / #184917

naveen-seth committed:
Merge branch 'main' into modules-driver-scan-and-graph

rnk reviewed (commented):
I don't feel sufficiently knowledgeable about C++ modules to really approve this, but I know this is an important change.

I went to the modules working group meeting a few months back, and this line of PRs to add dependency scanning to the clang driver was the main blocker for CMake C++ module support.

If we lack active maintainers willing to review modules patches, I just want to throw out the idea that we should empower @naveen-seth to be the new C++ module maintainer. If they are the resident expert, we should empower them. We want to bias to action. LLVM has sometimes been referred to as a do-aucracy, those who do the work get to have a say, and so far as I can tell, @naveen-seth is the one driving this forward.

The functionality also has very limited blast radius, since it's behind the not-widely-deployed `-fmodules-driver` flag.

ChuanqiXu9 commented:
> and this line of PRs to add dependency scanning to the clang driver was the main blocker for CMake C++ module support.

I didn't follow this. CMake already supports C++ modules. And I lack an interesting for this since I feel the feature is majorly used for using modules without a build system. But this is only common tutorial in my mind. @naveen-seth could you clarify?

> If we lack active maintainers willing to review modules patches, I just want to throw out the idea that we should empower @naveen-seth to be the new C++ module maintainer.

I don't mind this. In my mind, the more maintainer, the better.

Bigcheese reviewed (approved):
A have two minor comments, but I think this is in a state where it's good to land.

Bigcheese commented:
> Yes, the reasoning in that comment aligns with CMake's desires as well (particularly the second point). But yes, `clang-scan-deps` existing unblocked us and allowed for `clang` support to exist. Not having to `find_program` and hope it agrees with the `clang++` in use would remove a _lot_ of worry (of particular issue on macOS where Homebrew and Xcode toolchains can easily live in searched-by-default places).

At this point I think all the blockers on having `clang` do dep scanning are gone, we just need to commit to an interface. `clang-scan-deps` wasn't really intended to be anything but a test tool.

naveen-seth committed:
Address Bigcheese's review feedback

naveen-seth committed:
Fix formatting
